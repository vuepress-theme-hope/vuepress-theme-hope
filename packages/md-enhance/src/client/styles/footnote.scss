// footnote anchor position
.footnote-item {
  margin-top: calc(0rem - var(--header-offset, 3.6rem));
  padding-top: calc(var(--header-offset, 3.6rem) + 0.5rem);

  > p {
    margin-bottom: 0;
  }
}

.footnote-ref {
  position: relative;
}

.footnote-anchor {
  position: absolute;
  top: calc(-0.5rem - var(--header-offset, 3.6rem));
}
