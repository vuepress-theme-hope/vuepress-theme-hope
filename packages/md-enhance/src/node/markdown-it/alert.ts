import { alert as _alert } from "@mdit/plugin-alert";
import type { ExactLocaleConfig } from "@vuepress/helper";
import { ensureLeadingSlash } from "@vuepress/helper";
import type { PluginWithOptions } from "markdown-it";
import type { MarkdownEnv } from "vuepress/markdown";
import { resolveLocalePath } from "vuepress/shared";

import type { MarkdownAlertLocaleData } from "../typings/index.js";

export type MarkdownItAlertOptions = ExactLocaleConfig<MarkdownAlertLocaleData>;

export const alert: PluginWithOptions<MarkdownItAlertOptions> = (
  md,
  options = {},
) => {
  md.use(_alert, {
    alertNames: ["important", "note", "tip", "warning", "caution", "info"],
    openRender: (tokens, index): string =>
      `<div class="hint-container ${tokens[index].markup}">\n`,
    titleRender: (tokens, index, _options, env: MarkdownEnv) => {
      const type = tokens[index].markup;
      const relativePath = ensureLeadingSlash(env.filePathRelative ?? "");
      const localePath = resolveLocalePath(options, relativePath);

      return `\
<p class="hint-container-title">${
        options[localePath]?.[type as keyof MarkdownAlertLocaleData] || type
      }</p>
`;
    },
    closeRender: () => "</div>\n",
    deep: true,
  });
};
