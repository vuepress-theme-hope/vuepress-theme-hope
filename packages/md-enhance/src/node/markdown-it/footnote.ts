/**
 * Forked from https://github.com/markdown-it/markdown-it-footnote/blob/master/index.js
 *
 * Copyright (c) 2014-2015 Vitaly Puzrin, Alex Kocharin.
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

import { default as Token } from "markdown-it/lib/token.js";
import { default as parseLinkLabel } from "markdown-it/lib/helpers/parse_link_label.js";

import type { MarkdownEnv } from "@vuepress/markdown";
import type { default as Renderer } from "markdown-it/lib/renderer.js";
import type { default as StateBlock } from "markdown-it/lib/rules_block/state_block.js";
import type { default as StateCore } from "markdown-it/lib/rules_core/state_core.js";
import type { default as StateInline } from "markdown-it/lib/rules_inline/state_inline.js";
import type { PluginSimple } from "markdown-it";
import type { RuleBlock } from "markdown-it/lib/parser_block.js";
import type { RuleInline } from "markdown-it/lib/parser_inline.js";

interface FootNoteToken extends Token {
  meta: {
    id: number;
    subId: number;
    label: string;
  };
}

interface FootNoteEnv extends MarkdownEnv {
  docId?: string;
  footnotes: {
    label?: string;
    refs?: Record<string, number>;
    list?: {
      label?: string;
      count?: number;
      content?: string;
      tokens?: Token[] | null;
    }[];
  };
}

interface FootNoteStateBlock extends StateBlock {
  tokens: FootNoteToken[];
  env: FootNoteEnv;
}

interface FootNoteStateInline extends StateInline {
  tokens: FootNoteToken[];
  env: FootNoteEnv;
}

interface FootNoteStateCore extends StateCore {
  tokens: FootNoteToken[];
  env: FootNoteEnv;
}

const getIDSuffix = (tokens: FootNoteToken[], index: number): string =>
  // add suffix when multiple id was found
  tokens[index].meta.subId > 0 ? `:${tokens[index].meta.subId}` : "";

const renderFootnoteAnchorName: Renderer.RenderRule = (
  tokens: FootNoteToken[],
  index,
  _options,
  env: FootNoteEnv
): string =>
  `${
    // prefix
    typeof env.docId === "string" ? `-${env.docId}-` : ""
  }${
    // increasing id
    (tokens[index].meta.id + 1).toString()
  }`;

const renderFootnoteCaption: Renderer.RenderRule = (
  tokens: FootNoteToken[],
  index
): string =>
  `[${
    // number
    (tokens[index].meta.id + 1).toString()
  }${getIDSuffix(tokens, index)}]`;

const renderFootnoteRef: Renderer.RenderRule = (
  tokens: FootNoteToken[],
  index,
  options,
  env: FootNoteEnv,
  self
): string => {
  const id = self.rules["footnoteAnchorName"]!(
    tokens,
    index,
    options,
    env,
    self
  );
  const caption = self.rules["footnoteCaption"]!(
    tokens,
    index,
    options,
    env,
    self
  );

  return `<sup class="footnote-ref"><a href="#footnote${id}">${caption}</a><a class="footnote-anchor" id="footnote-ref${id}${getIDSuffix(
    tokens,
    index
  )}" /></sup>`;
};

const renderFootnoteBlockOpen: Renderer.RenderRule = (
  _tokens: FootNoteToken[],
  _index,
  options
): string =>
  `${
    options.xhtmlOut
      ? '<hr class="footnotes-sep" />\n'
      : '<hr class="footnotes-sep">\n'
  }<section class="footnotes">\n<ol class="footnotes-list">\n`;

const renderFootnoteBlockClose = (): string => "</ol>\n</section>\n";

const renderFootnoteOpen: Renderer.RenderRule = (
  tokens: FootNoteToken[],
  index,
  options,
  env: FootNoteEnv,
  self
): string =>
  `<li id="footnote${self.rules["footnoteAnchorName"]!(
    tokens,
    index,
    options,
    env,
    self
  )}${getIDSuffix(tokens, index)}" class="footnote-item">`;

const renderFootnoteClose = (): string => "</li>\n";

const renderFootnoteAnchor: Renderer.RenderRule = (
  tokens: FootNoteToken[],
  index,
  options,
  env: FootNoteEnv,
  self
): string => {
  return ` <a href="#footnote-ref${self.rules["footnoteAnchorName"]!(
    tokens,
    index,
    options,
    env,
    self
  )}${
    getIDSuffix(tokens, index)
    /* ↩ with escape code to prevent display as Apple Emoji on iOS */
  }" class="footnote-backref">\u21a9\uFE0E</a>`;
};

// Process footnote block definition
const footnoteDef: RuleBlock = (
  state: FootNoteStateBlock,
  startLine,
  endLine,
  silent
) => {
  let pos;
  let token;
  let offset;
  let ch;
  const start = state.bMarks[startLine] + state.tShift[startLine];
  const max = state.eMarks[startLine];

  // line should be at least 5 chars - "[^x]:"
  if (start + 4 > max) return false;

  if (state.src.charAt(start) !== "[") return false;

  if (state.src.charAt(start + 1) !== "^") return false;

  for (pos = start + 2; pos < max; pos++) {
    if (state.src.charAt(pos) === " ") return false;
    if (state.src.charAt(pos) === "]") break;
  }

  // no empty footnote labels
  if (pos === start + 2) return false;

  if (pos + 1 >= max || state.src.charAt(++pos) !== ":") return false;

  if (silent) return true;
  pos += 1;

  if (!state.env.footnotes) state.env.footnotes = {};
  if (!state.env.footnotes.refs) state.env.footnotes.refs = {};
  const label = state.src.slice(start + 2, pos - 2);

  state.env.footnotes.refs[`:${label}`] = -1;

  token = new Token("footnoteReferenceOpen", "", 1);
  token.meta = { label };
  token.level = state.level++;
  state.tokens.push(token);

  const oldBMark = state.bMarks[startLine];
  const oldTShift = state.tShift[startLine];
  const oldSCount = state.sCount[startLine];
  const oldParentType = state.parentType;
  const posAfterColon = pos;
  const initial =
    state.sCount[startLine] +
    pos -
    (state.bMarks[startLine] + state.tShift[startLine]);

  offset =
    state.sCount[startLine] +
    pos -
    (state.bMarks[startLine] + state.tShift[startLine]);

  while (pos < max) {
    ch = state.src.charAt(pos);

    if (ch === "\t") offset += 4 - (offset % 4);
    else if (ch === " ") offset += 1;
    else break;

    pos += 1;
  }

  state.tShift[startLine] = pos - posAfterColon;
  state.sCount[startLine] = offset - initial;

  state.bMarks[startLine] = posAfterColon;
  state.blkIndent += 4;
  state.parentType = "footnote" as unknown as StateBlock.ParentType;

  if (state.sCount[startLine] < state.blkIndent)
    state.sCount[startLine] += state.blkIndent;

  state.md.block.tokenize(state, startLine, endLine);

  state.parentType = oldParentType;
  state.blkIndent -= 4;
  state.tShift[startLine] = oldTShift;
  state.sCount[startLine] = oldSCount;
  state.bMarks[startLine] = oldBMark;

  token = new Token("footnoteReferenceClose", "", -1);
  token.level = --state.level;
  state.tokens.push(token);

  return true;
};

// Process inline footnotes (^[...])
const footnoteInline: RuleInline = (state: FootNoteStateInline, silent) => {
  let footnoteId;
  let token;
  let tokens: Token[];
  const max = state.posMax;
  const start = state.pos;

  if (start + 2 >= max) return false;
  if (state.src.charAt(start) !== "^") return false;

  if (state.src.charAt(start + 1) !== "[") return false;

  const labelStart = start + 2;
  const labelEnd = parseLinkLabel(state, start + 1);

  // parser failed to find ']', so it’s not a valid note
  if (labelEnd < 0) return false;

  /*
   * We found the end of the link, and know for a fact it’s a valid link;
   * so all that’s left to do is to call tokenizer.
   *
   */
  if (!silent) {
    if (!state.env.footnotes) state.env.footnotes = {};
    if (!state.env.footnotes.list) state.env.footnotes.list = [];
    footnoteId = state.env.footnotes.list.length;

    state.md.inline.parse(
      state.src.slice(labelStart, labelEnd),
      state.md,
      state.env,
      (tokens = [])
    );

    token = state.push("footnoteRef", "", 0);
    token.meta = { id: footnoteId };

    state.env.footnotes.list[footnoteId] = {
      content: state.src.slice(labelStart, labelEnd),
      tokens,
    };
  }

  state.pos = labelEnd + 1;
  state.posMax = max;

  return true;
};

// Process footnote references ([^...])
const footnoteRef: RuleInline = (state: FootNoteStateInline, silent) => {
  let pos;
  let footnoteId;
  let footnoteSubId;
  let token;
  const max = state.posMax;
  const start = state.pos;

  // should be at least 4 chars - "[^x]"
  if (start + 3 > max) return false;

  if (!state.env.footnotes || !state.env.footnotes.refs) return false;
  if (state.src.charAt(start) !== "[") return false;

  if (state.src.charAt(start + 1) !== "^") return false;

  for (pos = start + 2; pos < max; pos++) {
    if (state.src.charAt(pos) === " ") return false;
    if (state.src.charAt(pos) === "\n") return false;
    if (state.src.charAt(pos) === "]") break;
  }

  if (pos === start + 2) return false; // no empty footnote labels
  if (pos >= max) return false;

  pos += 1;

  const label = state.src.slice(start + 2, pos - 1);

  if (typeof state.env.footnotes.refs[`:${label}`] === "undefined")
    return false;

  if (!silent) {
    if (!state.env.footnotes.list) state.env.footnotes.list = [];

    if (state.env.footnotes.refs[`:${label}`] < 0) {
      footnoteId = state.env.footnotes.list.length;
      state.env.footnotes.list[footnoteId] = { label, count: 0 };
      state.env.footnotes.refs[`:${label}`] = footnoteId;
    } else footnoteId = state.env.footnotes.refs[`:${label}`];

    footnoteSubId = state.env.footnotes.list[footnoteId].count!;
    state.env.footnotes.list[footnoteId].count =
      state.env.footnotes.list[footnoteId].count! + 1;

    token = state.push("footnoteRef", "", 0);
    token.meta = { id: footnoteId, subId: footnoteSubId, label };
  }

  state.pos = pos;
  state.posMax = max;

  return true;
};

// Glue footnote tokens to end of token stream
const footnoteTail = (state: FootNoteStateCore): boolean => {
  let lastParagraph: FootNoteToken | null;
  let token: Token;
  let tokens: Token[];
  let current: Token[];
  let currentLabel: string;
  let insideRef = false;
  const refTokens: Record<string, Token[]> = {};

  if (!state.env.footnotes) return false;

  state.tokens = state.tokens.filter((stateToken) => {
    if (stateToken.type === "footnoteReferenceOpen") {
      insideRef = true;
      current = [];
      currentLabel = stateToken.meta.label;

      return false;
    }
    if (stateToken.type === "footnoteReferenceClose") {
      insideRef = false;
      // prepend ':' to avoid conflict with Object.prototype members
      refTokens[`:${currentLabel}`] = current;

      return false;
    }
    if (insideRef) current.push(stateToken);

    return !insideRef;
  });

  if (!state.env.footnotes.list) return false;
  const { list } = state.env.footnotes;

  token = new Token("footnoteBlockOpen", "", 1);
  state.tokens.push(token);

  for (let i = 0, { length } = list; i < length; i++) {
    token = new Token("footnoteOpen", "", 1);
    token.meta = { id: i, label: list[i].label };
    state.tokens.push(token);

    if (list[i].tokens) {
      tokens = [];

      token = new Token("paragraph_open", "p", 1);
      token.block = true;
      tokens.push(token);

      token = new Token("inline", "", 0);
      token.children = list[i].tokens!;
      token.content = list[i].content!;
      tokens.push(token);

      token = new Token("paragraph_close", "p", -1);
      token.block = true;
      tokens.push(token);
    } else if (list[i].label) tokens = refTokens[`:${list[i].label!}`];
    else tokens = [];

    if (tokens) state.tokens = state.tokens.concat(tokens);
    if (state.tokens[state.tokens.length - 1].type === "paragraph_close")
      lastParagraph = state.tokens.pop() || null;
    else lastParagraph = null;

    for (let j = 0; j < (Number(list[i].count) > 0 ? list[i].count! : 1); j++) {
      token = new Token("footnoteAnchor", "", 0);
      token.meta = { id: i, subId: j, label: list[i].label };
      state.tokens.push(token);
    }

    if (lastParagraph) state.tokens.push(lastParagraph);

    token = new Token("footnoteClose", "", -1);
    state.tokens.push(token);
  }

  token = new Token("footnoteBlockClose", "", -1);
  state.tokens.push(token);

  return true;
};

export const footnote: PluginSimple = (md) => {
  md.renderer.rules["footnoteRef"] = renderFootnoteRef;
  md.renderer.rules["footnoteBlockOpen"] = renderFootnoteBlockOpen;
  md.renderer.rules["footnoteBlockClose"] = renderFootnoteBlockClose;
  md.renderer.rules["footnoteOpen"] = renderFootnoteOpen;
  md.renderer.rules["footnoteClose"] = renderFootnoteClose;
  md.renderer.rules["footnoteAnchor"] = renderFootnoteAnchor;

  // helpers (only used in other rules, no tokens are attached to those)
  // helpers (only used in other rules, no tokens are attached to those)
  md.renderer.rules["footnoteCaption"] = renderFootnoteCaption;
  md.renderer.rules["footnoteAnchorName"] = renderFootnoteAnchorName;

  md.block.ruler.before("reference", "footnoteDef", footnoteDef, {
    alt: ["paragraph", "reference"],
  });
  md.inline.ruler.after("image", "footnoteInline", footnoteInline);
  md.inline.ruler.after("footnoteInline", "footnoteRef", footnoteRef);
  md.core.ruler.after("inline", "footnoteTail", footnoteTail);
};
