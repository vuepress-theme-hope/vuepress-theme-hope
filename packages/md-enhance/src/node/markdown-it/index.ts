export * from "./align.js";
export { attrs } from "./attrs.js";
export * from "./chart.js";
export * from "./codeDemo.js";
export * from "./codeTabs.js";
export * from "./container.js";
export * from "./echarts.js";
export * from "./flowchart.js";
export * from "./footnote.js";
export * from "./hint.js";
export * from "./imageMark.js";
export * from "./imageSize.js";
export * from "./katex.js";
export * from "./mathjax.js";
export * from "./lazyLoad.js";
export * from "./mark.js";
export * from "./include.js";
export * from "./mermaid.js";
export * from "./playground/index.js";
export * from "./presentation.js";
export * from "./sub.js";
export * from "./sup.js";
export * from "./tabs.js";
export * from "./tasklist.js";
export * from "./tex.js";
export * from "./vPre.js";
export * from "./stylize.js";
export * from "./uml.js";
export * from "./vuePlayground.js";
