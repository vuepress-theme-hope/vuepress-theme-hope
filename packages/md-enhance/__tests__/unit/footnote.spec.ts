import { describe, it, expect } from "vitest";
import MarkdownIt from "markdown-it";
import { footnote } from "../../src/node/markdown-it/footnote";

describe("footnote", () => {
  const markdownIt = MarkdownIt({ linkify: true }).use(footnote);

  it("Pandoc example", () => {
    expect(
      markdownIt.render(
        `
Here is a footnote reference,[^1] and another.[^longnote]

[^1]: Here is the footnote.

[^longnote]: Here's one with multiple blocks.

    Subsequent paragraphs are indented to show that they
belong to the previous footnote.

        { some.code }

    The whole paragraph can be indented, or just the first
    line.  In this way, multi-paragraph footnotes work like
    multi-paragraph list items.

This paragraph won’t be part of the note, because it
isn't indented.
`,
        {}
      )
    ).toMatchSnapshot();
  });

  it("Should terminate each other", () => {
    expect(
      markdownIt.render(`
[^1][^2][^3]

[^1]: foo
[^2]: bar
[^3]: baz
`)
    ).toMatchSnapshot();
  });

  it("Can inside blockquotes, and are lazy", () => {
    expect(
      markdownIt.render(`
[^foo]

> [^foo]: bar
baz
`)
    ).toMatchSnapshot();
  });

  it("Label cannot contain spaces or newlines", () => {
    expect(
      markdownIt.render(`
[^ foo]: bar baz

[^foo
]: bar baz
`)
    ).toMatchSnapshot();
  });

  it("Nested footnotes", () => {
    expect(
      markdownIt.render(`
foo[^1] bar[^2].

[^1]:[^2]: baz
`)
    ).toMatchSnapshot();
  });

  it("Should support inline note", () => {
    expect(
      markdownIt.render(`
Here is an inline note.^[Inlines notes are easier to write, since
you don’t have to pick an identifier and move down to type the
note.]
`)
    ).toMatchSnapshot();
  });

  it("Can have arbitrary markup", () => {
    expect(markdownIt.render(`foo^[ *bar* ]`)).toMatchSnapshot();
  });

  it("Duplicate footnotes shoud have suffix", () => {
    expect(
      markdownIt.render(`
[^xxxxx] [^xxxxx]

[^xxxxx]: foo
`)
    ).toMatchSnapshot();
  });

  it("Indent", () => {
    expect(
      markdownIt.render(`
[^xxxxx] [^yyyyy]

[^xxxxx]: foo
    ---

[^yyyyy]: foo
    ---
`)
    ).toMatchSnapshot();
  });

  it("Indents for the first line", () => {
    expect(
      markdownIt.render(`
[^xxxxx] [^yyyyy]

[^xxxxx]:       foo

[^yyyyy]:        foo
`)
    ).toMatchSnapshot();

    expect(
      markdownIt.render(`
[^xxxxx]

[^xxxxx]:		foo
`)
    ).toMatchSnapshot();
  });

  it("Should contain Security", () => {
    expect(
      markdownIt.render(`
[^__proto__]

[^__proto__]: blah
`)
    ).toMatchSnapshot();

    expect(
      markdownIt.render(`
[^hasOwnProperty]

[^hasOwnProperty]: blah
`)
    ).toMatchSnapshot();
  });

  it("Should allow links in inline footnotes", () => {
    expect(
      markdownIt.render(`Example^[this is another example https://github.com]`)
    ).toMatchSnapshot();
  });

  it("custom docId in env", () => {
    expect(
      markdownIt.render(
        `
Here is a footnote reference,[^1] and another.[^longnote]

[^1]: Here is the footnote.

[^longnote]: Here's one with multiple blocks.

    Subsequent paragraphs are indented to show that they
belong to the previous footnote.

        { some.code }

    The whole paragraph can be indented, or just the first
    line.  In this way, multi-paragraph footnotes work like
    multi-paragraph list items.

This paragraph won’t be part of the note, because it
isn't indented.
`,
        { docId: "test-doc-id" }
      )
    ).toMatchSnapshot();
  });
});
