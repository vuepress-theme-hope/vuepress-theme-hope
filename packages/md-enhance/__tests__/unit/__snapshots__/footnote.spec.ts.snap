// Vitest Snapshot v1

exports[`footnote > Can have arbitrary markup 1`] = `
"<p>foo<sup class=\\"footnote-ref\\"><a href=\\"#footnote1\\">[1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref1\\" /></sup></p>
<hr class=\\"footnotes-sep\\">
<section class=\\"footnotes\\">
<ol class=\\"footnotes-list\\">
<li id=\\"footnote1\\" class=\\"footnote-item\\"><p> <em>bar</em>  <a href=\\"#footnote-ref1\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
</ol>
</section>
"
`;

exports[`footnote > Can inside blockquotes, and are lazy 1`] = `
"<p><sup class=\\"footnote-ref\\"><a href=\\"#footnote1\\">[1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref1\\" /></sup></p>
<blockquote></blockquote>
<hr class=\\"footnotes-sep\\">
<section class=\\"footnotes\\">
<ol class=\\"footnotes-list\\">
<li id=\\"footnote1\\" class=\\"footnote-item\\"><p>bar
baz <a href=\\"#footnote-ref1\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
</ol>
</section>
"
`;

exports[`footnote > Duplicate footnotes shoud have suffix 1`] = `
"<p><sup class=\\"footnote-ref\\"><a href=\\"#footnote1\\">[1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref1\\" /></sup> <sup class=\\"footnote-ref\\"><a href=\\"#footnote1\\">[1:1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref1:1\\" /></sup></p>
<hr class=\\"footnotes-sep\\">
<section class=\\"footnotes\\">
<ol class=\\"footnotes-list\\">
<li id=\\"footnote1\\" class=\\"footnote-item\\"><p>foo <a href=\\"#footnote-ref1\\" class=\\"footnote-backref\\">↩︎</a> <a href=\\"#footnote-ref1:1\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
</ol>
</section>
"
`;

exports[`footnote > Indent 1`] = `
"<p><sup class=\\"footnote-ref\\"><a href=\\"#footnote1\\">[1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref1\\" /></sup> <sup class=\\"footnote-ref\\"><a href=\\"#footnote2\\">[2]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref2\\" /></sup></p>
<hr class=\\"footnotes-sep\\">
<section class=\\"footnotes\\">
<ol class=\\"footnotes-list\\">
<li id=\\"footnote1\\" class=\\"footnote-item\\"><h2>foo</h2>
 <a href=\\"#footnote-ref1\\" class=\\"footnote-backref\\">↩︎</a></li>
<li id=\\"footnote2\\" class=\\"footnote-item\\"><h2>foo</h2>
 <a href=\\"#footnote-ref2\\" class=\\"footnote-backref\\">↩︎</a></li>
</ol>
</section>
"
`;

exports[`footnote > Indents for the first line 1`] = `
"<p><sup class=\\"footnote-ref\\"><a href=\\"#footnote1\\">[1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref1\\" /></sup> <sup class=\\"footnote-ref\\"><a href=\\"#footnote2\\">[2]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref2\\" /></sup></p>
<hr class=\\"footnotes-sep\\">
<section class=\\"footnotes\\">
<ol class=\\"footnotes-list\\">
<li id=\\"footnote1\\" class=\\"footnote-item\\"><p>foo <a href=\\"#footnote-ref1\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
<li id=\\"footnote2\\" class=\\"footnote-item\\"><pre><code>foo
</code></pre>
 <a href=\\"#footnote-ref2\\" class=\\"footnote-backref\\">↩︎</a></li>
</ol>
</section>
"
`;

exports[`footnote > Indents for the first line 2`] = `
"<p><sup class=\\"footnote-ref\\"><a href=\\"#footnote1\\">[1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref1\\" /></sup></p>
<hr class=\\"footnotes-sep\\">
<section class=\\"footnotes\\">
<ol class=\\"footnotes-list\\">
<li id=\\"footnote1\\" class=\\"footnote-item\\"><p>foo <a href=\\"#footnote-ref1\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
</ol>
</section>
"
`;

exports[`footnote > Label cannot contain spaces or newlines 1`] = `
"<p>[^ foo]: bar baz</p>
<p>[^foo
]: bar baz</p>
"
`;

exports[`footnote > Nested footnotes 1`] = `
"<p>foo<sup class=\\"footnote-ref\\"><a href=\\"#footnote1\\">[1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref1\\" /></sup> bar<sup class=\\"footnote-ref\\"><a href=\\"#footnote2\\">[2]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref2\\" /></sup>.</p>
<hr class=\\"footnotes-sep\\">
<section class=\\"footnotes\\">
<ol class=\\"footnotes-list\\">
<li id=\\"footnote1\\" class=\\"footnote-item\\"> <a href=\\"#footnote-ref1\\" class=\\"footnote-backref\\">↩︎</a></li>
<li id=\\"footnote2\\" class=\\"footnote-item\\"><p>baz <a href=\\"#footnote-ref2\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
</ol>
</section>
"
`;

exports[`footnote > Pandoc example 1`] = `
"<p>Here is a footnote reference,<sup class=\\"footnote-ref\\"><a href=\\"#footnote1\\">[1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref1\\" /></sup> and another.<sup class=\\"footnote-ref\\"><a href=\\"#footnote2\\">[2]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref2\\" /></sup></p>
<p>This paragraph won’t be part of the note, because it
isn't indented.</p>
<hr class=\\"footnotes-sep\\">
<section class=\\"footnotes\\">
<ol class=\\"footnotes-list\\">
<li id=\\"footnote1\\" class=\\"footnote-item\\"><p>Here is the footnote. <a href=\\"#footnote-ref1\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
<li id=\\"footnote2\\" class=\\"footnote-item\\"><p>Here's one with multiple blocks.</p>
<p>Subsequent paragraphs are indented to show that they
belong to the previous footnote.</p>
<pre><code>{ some.code }
</code></pre>
<p>The whole paragraph can be indented, or just the first
line.  In this way, multi-paragraph footnotes work like
multi-paragraph list items. <a href=\\"#footnote-ref2\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
</ol>
</section>
"
`;

exports[`footnote > Should allow links in inline footnotes 1`] = `
"<p>Example<sup class=\\"footnote-ref\\"><a href=\\"#footnote1\\">[1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref1\\" /></sup></p>
<hr class=\\"footnotes-sep\\">
<section class=\\"footnotes\\">
<ol class=\\"footnotes-list\\">
<li id=\\"footnote1\\" class=\\"footnote-item\\"><p>this is another example <a href=\\"https://github.com\\">https://github.com</a> <a href=\\"#footnote-ref1\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
</ol>
</section>
"
`;

exports[`footnote > Should contain Security 1`] = `
"<p><sup class=\\"footnote-ref\\"><a href=\\"#footnote1\\">[1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref1\\" /></sup></p>
<hr class=\\"footnotes-sep\\">
<section class=\\"footnotes\\">
<ol class=\\"footnotes-list\\">
<li id=\\"footnote1\\" class=\\"footnote-item\\"><p>blah <a href=\\"#footnote-ref1\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
</ol>
</section>
"
`;

exports[`footnote > Should contain Security 2`] = `
"<p><sup class=\\"footnote-ref\\"><a href=\\"#footnote1\\">[1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref1\\" /></sup></p>
<hr class=\\"footnotes-sep\\">
<section class=\\"footnotes\\">
<ol class=\\"footnotes-list\\">
<li id=\\"footnote1\\" class=\\"footnote-item\\"><p>blah <a href=\\"#footnote-ref1\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
</ol>
</section>
"
`;

exports[`footnote > Should support inline note 1`] = `
"<p>Here is an inline note.<sup class=\\"footnote-ref\\"><a href=\\"#footnote1\\">[1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref1\\" /></sup></p>
<hr class=\\"footnotes-sep\\">
<section class=\\"footnotes\\">
<ol class=\\"footnotes-list\\">
<li id=\\"footnote1\\" class=\\"footnote-item\\"><p>Inlines notes are easier to write, since
you don’t have to pick an identifier and move down to type the
note. <a href=\\"#footnote-ref1\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
</ol>
</section>
"
`;

exports[`footnote > Should terminate each other 1`] = `
"<p><sup class=\\"footnote-ref\\"><a href=\\"#footnote1\\">[1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref1\\" /></sup><sup class=\\"footnote-ref\\"><a href=\\"#footnote2\\">[2]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref2\\" /></sup><sup class=\\"footnote-ref\\"><a href=\\"#footnote3\\">[3]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref3\\" /></sup></p>
<hr class=\\"footnotes-sep\\">
<section class=\\"footnotes\\">
<ol class=\\"footnotes-list\\">
<li id=\\"footnote1\\" class=\\"footnote-item\\"><p>foo <a href=\\"#footnote-ref1\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
<li id=\\"footnote2\\" class=\\"footnote-item\\"><p>bar <a href=\\"#footnote-ref2\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
<li id=\\"footnote3\\" class=\\"footnote-item\\"><p>baz <a href=\\"#footnote-ref3\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
</ol>
</section>
"
`;

exports[`footnote > custom docId in env 1`] = `
"<p>Here is a footnote reference,<sup class=\\"footnote-ref\\"><a href=\\"#footnote-test-doc-id-1\\">[1]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref-test-doc-id-1\\" /></sup> and another.<sup class=\\"footnote-ref\\"><a href=\\"#footnote-test-doc-id-2\\">[2]</a><a class=\\"footnote-anchor\\" id=\\"footnote-ref-test-doc-id-2\\" /></sup></p>
<p>This paragraph won’t be part of the note, because it
isn't indented.</p>
<hr class=\\"footnotes-sep\\">
<section class=\\"footnotes\\">
<ol class=\\"footnotes-list\\">
<li id=\\"footnote-test-doc-id-1\\" class=\\"footnote-item\\"><p>Here is the footnote. <a href=\\"#footnote-ref-test-doc-id-1\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
<li id=\\"footnote-test-doc-id-2\\" class=\\"footnote-item\\"><p>Here's one with multiple blocks.</p>
<p>Subsequent paragraphs are indented to show that they
belong to the previous footnote.</p>
<pre><code>{ some.code }
</code></pre>
<p>The whole paragraph can be indented, or just the first
line.  In this way, multi-paragraph footnotes work like
multi-paragraph list items. <a href=\\"#footnote-ref-test-doc-id-2\\" class=\\"footnote-backref\\">↩︎</a></p>
</li>
</ol>
</section>
"
`;
